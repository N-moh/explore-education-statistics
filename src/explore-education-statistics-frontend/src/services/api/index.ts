import axios from 'axios';
import { commaSeparated } from '../util/paramSerializers';

export const baseUrl = {
  content: process.env.CONTENT_API_BASE_URL,
  data: process.env.DATA_API_BASE_URL,
};

export const contentApi = axios.create({
  baseURL: `${baseUrl.content}/api/`,
  paramsSerializer: commaSeparated,
});

export const dataApi = axios.create({
  baseURL: `${baseUrl.data}/api/`,
  paramsSerializer: commaSeparated,
});
