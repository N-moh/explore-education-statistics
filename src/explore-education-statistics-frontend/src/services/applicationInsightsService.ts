import { ApplicationInsights } from '@microsoft/applicationinsights-web';

// eslint-disable-next-line import/prefer-default-export
export const initApplicationInsights = () => {
  if (
    process.env.AI_INSTRUMENTATION !== undefined &&
    process.env.AI_TRACKING === 'true'
  ) {
    const appInsights = new ApplicationInsights({
      config: {
        instrumentationKey: process.env.AI_INSTRUMENTATION,
        enableAutoRouteTracking: true,
        autoTrackPageVisitTime: true,
      },
    });
    appInsights.loadAppInsights();

    // eslint-disable-next-line no-console
    console.log('Application Insights initialised');
  }
};
