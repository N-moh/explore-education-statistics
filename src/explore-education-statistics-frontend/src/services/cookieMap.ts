function daysFromNow(days = 0) {
  const result = new Date();
  if (!days) {
    // Add one month
    result.setMonth([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 0][result.getMonth()]);
  } else {
    result.setDate(result.getDate() + days);
  }
  return result;
}

const CookieMap = {
  bannerSeenCookie: {
    name: 'ees_banner_seen',
    expires: daysFromNow(),
    duration: '1 month',
  },
  disableGACookie: {
    name: 'ees_disable_google_analytics',
    expires: daysFromNow(),
    duration: '1 month',
  },
};

export default CookieMap;
