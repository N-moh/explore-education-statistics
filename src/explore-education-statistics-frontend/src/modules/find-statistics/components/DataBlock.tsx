/* eslint-disable */
import Tabs from '@common/components/Tabs';
import TabsSection from '@common/components/TabsSection';
import { MapFeature } from '@common/modules/find-statistics/components/charts/MapBlock';
import SummaryRenderer, {
  SummaryRendererProps,
} from '@common/modules/find-statistics/components/SummaryRenderer';
import TableRenderer2, {
  Props as TableRendererProps,
} from '@common/modules/find-statistics/components/TableRenderer2';
import { baseUrl } from '@common/services/api';
import { Chart, DataQuery, Summary } from '@common/services/publicationService';
import {
  CharacteristicsData,
  PublicationMeta,
} from '@common/services/tableBuilderService';
import Link from '@frontend/components/Link';
import React, { Component } from 'react';
import DataBlockService, {
  DataBlockRequest,
  DataBlockMetadata,
  DataBlockData,
} from '@common/services/dataBlockService';
import {
  ChartRenderer,
  ChartRendererProps,
} from '@common/modules/find-statistics/components/ChartRenderer';

export interface DataBlockProps {
  type: string;
  heading?: string;
  dataBlockRequest?: DataBlockRequest;
  charts?: Chart[];
  summary?: Summary;
  data?: CharacteristicsData;
  meta?: PublicationMeta;
  height?: number;
  showTables?: boolean;
}

interface DataBlockState {
  charts?: ChartRendererProps[];
  // downloads?: any[];
  tables?: TableRendererProps[];
  summary?: SummaryRendererProps;
}

class DataBlock extends Component<DataBlockProps, DataBlockState> {
  public static defaultProps = {
    showTables: true,
  };

  public state: DataBlockState = {};

  private currentDataQuery?: DataQuery = undefined;

  public async componentDidMount() {
    const { dataBlockRequest, data, meta } = this.props;

    if (dataBlockRequest) {
      const result = await DataBlockService.getDataBlockForSubject(
        dataBlockRequest,
      );

      this.parseDataResponse(result.data, result.metaData);

      console.log(dataBlockRequest);
    } else {
      // this.parseDataResponse(data, meta);
    }
  }

  public async componentWillUnmount() {
    this.currentDataQuery = undefined;
  }

  private parseDataResponse(
    json: DataBlockData,
    jsonMeta: DataBlockMetadata,
  ): void {
    const newState: DataBlockState = {};

    if (json.result.length > 0) {
      newState.tables = [{ data: json, meta: jsonMeta }];
    }

    const { charts, summary } = this.props;

    if (charts) {
      newState.charts = charts.map(chart => ({
        ...chart,
        geometry: chart.geometry as MapFeature,
        data: json,
        meta: jsonMeta,
      }));
    }

    this.setState(newState);
  }

  public render() {
    const id = new Date().getDate();

    const { heading, height, showTables } = this.props;
    const { charts, summary, tables } = this.state;

    return (
      <div className="govuk-datablock" data-testid={`DataBlock ${heading}`}>
        <Tabs>
          {summary && (
            <TabsSection id={`${id}_summary`} title="Summary">
              <h3>{heading}</h3>
              <SummaryRenderer {...summary} />
            </TabsSection>
          )}

          {tables && showTables && (
            <TabsSection id={`${id}0`} title="Data tables">
              <h3>{heading}</h3>
              {tables.map((table, idx) => {
                const key = `${id}0_table_${idx}`;

                return <TableRenderer2 key={key} {...table} />;
              })}
              <h2 className="govuk-heading-m govuk-!-margin-top-9">
                Explore and edit this data online
              </h2>
              <p>Use our table tool to add and remove data for this table.</p>
              <Link to="/table-tool/" className="govuk-button">
                Explore data
              </Link>
            </TabsSection>
          )}

          {charts && (
            <TabsSection id={`${id}1`} title="Charts" lazy={showTables}>
              <h3>{heading}</h3>
              {charts.map((chart, idx) => {
                const key = `${id}_chart_${idx}`;

                return <ChartRenderer key={key} {...chart} height={height} />;
              })}
              <h2 className="govuk-heading-m govuk-!-margin-top-9">
                Explore and edit this data online
              </h2>
              <p>Use our table tool to add and remove data for this chart.</p>
              <Link to="/table-tool/" className="govuk-button">
                Explore chart data
              </Link>
            </TabsSection>
          )}

          <TabsSection id={`${id}2`} title="Data downloads">
            <p>
              You can customise and download data as Excel or .csv files. Our
              data can also be accessed via an API.
            </p>
            <div className="govuk-inset-text">
              Data downloads have not yet been implemented within the service.
            </div>
            <h2 className="govuk-heading-m govuk-!-margin-top-9">
              Explore and edit this data online
            </h2>
            <p>Use our table tool to explore this data.</p>
            <Link to="/table-tool/" className="govuk-button">
              Explore data
            </Link>
          </TabsSection>
        </Tabs>
      </div>
    );
  }
}

export default DataBlock;
