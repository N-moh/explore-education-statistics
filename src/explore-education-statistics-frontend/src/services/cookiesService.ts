import * as googleAnalyticsService from '@frontend/services/googleAnalyticsService';

interface CookieObj {
  name: string;
  value?: any;
  expires?: Date;
}

function oneMonthFromNow() {
  const dateToday = new Date();
  return new Date(dateToday.setMonth(dateToday.getMonth() + 1));
}

export const bannerSeenCookieName = 'ees_banner_seen';
export const cookieSettingsCookieName = 'ees_cookie_settings';

export function getCookies() {
  // turns document.cookie e.g. 'cookie_name=hello-world; cookietwo=true'
  // into { cookie_name: "hellow-world", cookietwo: true }
  const cookies: any = {};
  document.cookie.split('; ').forEach(cookiesPairString => {
    const [key, value]: string[] = cookiesPairString.split('=');
    cookies[key] = value;
  });
  return cookies;
}

export function setCookies(newCookies: CookieObj[]) {
  newCookies.forEach(cookie => {
    const { name, value, expires = oneMonthFromNow() } = cookie;
    document.cookie = `${name}=${JSON.stringify(value)}; expires=${expires}`;

    if (name === cookieSettingsCookieName) {
      const cookieSettingGA = value.googleAnalytics;
      if (cookieSettingGA && cookieSettingGA.length) {
        if (cookieSettingGA === 'off') {
          googleAnalyticsService.disableGA();
        } else {
          googleAnalyticsService.enableGA();
        }
      }
    }
  });
}

export function deleteCookies(cookieNames: string[]) {
  setCookies(
    cookieNames.map(cookieName => {
      return {
        name: cookieName,
        expires: new Date(0),
      };
    }),
  );
}

export function acceptCookies() {
  // Accepted cookies banner
  setCookies([
    {
      name: bannerSeenCookieName,
      value: 'true',
    },
  ]);
}
