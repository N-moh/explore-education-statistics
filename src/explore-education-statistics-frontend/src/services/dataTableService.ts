import { AxiosPromise } from 'axios';
import { dataApi } from './api';

export enum SchoolType {
  Dummy = 'Dummy',
  Total = 'Total',
  State_Funded_Primary = 'State_Funded_Primary',
  State_Funded_Secondary = 'State_Funded_Secondary',
  Special = 'Special',
}

export interface DataTableResult {
  year: number;
  schoolType: SchoolType;
  attributes: {
    [attribute: string]: string;
  };
}

export interface CharacteristicsData {
  publicationId: string;
  releaseId: string;
  releaseDate: string;
  result: DataTableResult[];
}

export const getNationalCharacteristicsData = (
  publicationUuid: string,
  attributes: string[],
  characteristics: string[],
  schoolTypes: SchoolType[],
  years?: number[],
): AxiosPromise<CharacteristicsData> =>
  dataApi.get(`/tablebuilder/characteristics/national/${publicationUuid}`, {
    params: {
      attribute: attributes,
      characteristic: characteristics,
      schoolType: schoolTypes,
      year: years,
    },
  });
