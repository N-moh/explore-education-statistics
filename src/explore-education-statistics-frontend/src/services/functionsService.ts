import { functionApi } from '@common/services/api';

export interface SubscriptionData {
  email: string;
  slug: string;
  title: string;
}

export default {
  subscribeToPublication(query: {
    email: string;
    id: string;
    slug: string;
    title: string;
  }): Promise<SubscriptionData> {
    return functionApi.post('/publication/subscribe', query);
  },
};
