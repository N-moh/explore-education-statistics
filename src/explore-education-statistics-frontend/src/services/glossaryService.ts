import { contentApi } from '@common/services/api';
import {
  GlossaryCategory,
  GlossaryEntry,
} from '@common/services/types/glossary';

const glossaryService = {
  listGlossaryEntries(): Promise<GlossaryCategory[]> {
    return contentApi.get('/glossary', {});
  },
  getGlossaryEntry(slug: string): Promise<GlossaryEntry> {
    return contentApi.get(`glossary/${slug}`, {});
  },
};

export default glossaryService;
