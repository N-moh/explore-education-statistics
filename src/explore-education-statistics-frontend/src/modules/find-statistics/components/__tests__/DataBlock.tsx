import React from 'react';
import { render } from 'react-testing-library';
import { GeographicLevel } from 'explore-education-statistics-common/src/services/dataBlockService';
import DataBlock from '../DataBlock';

describe('DataBlock', () => {
  test('datablock test', () => {
    const { container } = render(
      <DataBlock
        type="datablock"
        dataQuery={{
          subjectId: 1,
          geographicLevel: GeographicLevel.National,
          startYear: 2014,
          endYear: 2015,
          filters: [1],
          indicators: [23, 26, 28],
        }}
      />,
    );
    expect(container.innerHTML).toMatchSnapshot();
  });
});
