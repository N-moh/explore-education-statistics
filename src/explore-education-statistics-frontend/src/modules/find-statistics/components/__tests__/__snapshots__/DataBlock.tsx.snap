// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataBlock datablock test 1`] = `

<div class="govuk-datablock"
     data-testid="DataBlock undefined"
>
  <div class="govuk-tabs">
    <ul class="govuk-tabs__list"
        role="tablist"
    >
      <li class="govuk-tabs__list-item"
          role="presentation"
      >
        <a class="govuk-tabs__tab govuk-tabs__tab--selected"
           href="#22"
           id="22-tab"
           aria-controls="22"
           aria-selected="true"
           role="tab"
        >
          Data downloads
        </a>
      </li>
    </ul>
    <section class="govuk-tabs__panel panel"
             id="22"
             data-testid="Data downloads"
             aria-labelledby="22-tab"
             role="tabpanel"
             tabindex="-1"
    >
      <h3>
        Data downloads
      </h3>
      <p>
        You can customise and download data as Excel or .csv files. Our data can also be accessed via an API.
      </p>
      <div class="govuk-inset-text">
        Data downloads have not yet been implemented within the service.
      </div>
      <h2 class="govuk-heading-m govuk-!-margin-top-9">
        Explore and edit this data online
      </h2>
      <p>
        Use our table tool to explore this data.
      </p>
      <a class="govuk-link govuk-button"
         href="/table-tool/"
      >
        Explore data
      </a>
    </section>
  </div>
</div>

`;
